src/crypto/tls/handshake_client.go

```diff
@@ -838,7 +950,7 @@ func (c *Conn) verifyServerCertificate(certificates [][]byte) error {
 		opts := x509.VerifyOptions{
 			Roots:         c.config.RootCAs,
 			CurrentTime:   c.config.time(),
-			DNSName:       c.config.ServerName,
+			DNSName:       c.serverName,
```

Comment:
I don't understand. The name that's verified is the inner SNI when ECH accepted and outer SNI otherwise. `c.serverName` is set to whatever name is verified: `c.config.ServerName` is only the correct name in case ECH is accepted.

What's weird is that I reverted this change and the tests still passed. Maybe I'm misunderstanding the meaning of this field?

Reply:
Perhaps tests are being executed with InsecureSkipVerify set?

`hello.serverName` is not the same as `config.ServerName`. The latter is derived from the former which can be empty if the `config.ServerName` is an IP address such as `1.1.1.1`. If `DNSName` is empty, then no hostname validation is performed, allowing any certificate signed by a valid CA to be accepted for requests such as `http.Get("https://1.1.1.1/`")`.

References:
- https://golang.org/pkg/crypto/x509/#VerifyOptions.DNSName
- https://golang.org/pkg/crypto/tls/#Config.ServerName

Reply:
@cjpatton 
> What's weird is that I reverted this change and the tests still passed.

After investigation, I found that there was a bug in the test code that made it appear occasionally that there was no error when in fact there was one. This has been fixed, and I can confirm that the code is correct as-is. (If you revert this to `c.config.ServerName`, several tests will fail.) 

Let me explain. `c.serverName` is set to whatever server name the client will verify. This is `c.config.ServerName` if ECH is accepted or bypassed and `echConfig.PublicName()` if ECH is rejected. (`echConfig` denotes the ECH config selected by the client.)

Reply:
I still think this is wrong, `c.serverName = hello.serverName` fills in the server name from the Client Hello. If the client configured ServerName as 1.1.1.1 for example, then the client hello will not have SNI.

I believe that with this change, certificate validation is disabled even when ECH is not enabled. I could be mistaken, but I am pretty sure this is currently the case with the proposed change. This is a blocker for merging this change. Could you convince me that the code does not have this issue?

Reply:
I don't agree that validation is being disabled by this change, as the tests don't bare this out. In any case, I changed the code to something you might prefer. `opts` is now defined as follows:
```language: go
        dnsName := c.config.ServerName
        if c.ech.st.Offered && !c.ech.st.Accepted {
            dnsName = c.serverName
        }   
        opts := x509.VerifyOptions{                                                                                                                                                                                                                                             
            Roots:         c.config.RootCAs,
            CurrentTime:   c.config.time(),
            DNSName:       dnsName,
            Intermediates: x509.NewCertPool(),
        }   
```